core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -6542,6 +6546,27 @@ sealed trait Exit[+E, +A] extends ZIO[Any, E, A] { self =>
       case (e @ Failure(_), _)      => e
       case (_, e @ Failure(_))      => e
     }
+
+  private def zipLeftWith[E1 >: E](that: Exit[E1, Any])(
+    g: (Cause[E], Cause[E1]) => Cause[E1]
+  ): Exit[E1, A] =
+    (self, that) match {
+      case (l @ Success(_), Success(_)) => l
+      case (Failure(l), Failure(r))     => Exit.failCause(g(l, r))
+      case (e @ Failure(_), _)          => e
+      case (_, e @ Failure(_))          => e
+    }
+
+  private def zipRightWith[E1 >: E, B](that: Exit[E1, B])(
+    g: (Cause[E], Cause[E1]) => Cause[E1]
+  ): Exit[E1, B] =
+    (self, that) match {
```

Comment:
this allocates a tuple - won't a nested pattern match be better?

Reply:
John made me change this kind of pattern matching to something like:
```scala
self match {
  case ... => that match {
     case => ...
  }
}
```
for this reason. See https://github.com/zio/zio/pull/8746#discussion_r1567448064

Reply:
Damn, I would have expected the Scala compiler to optimize this. Oh well, I'll change it, thanks for the heads up 👍